File: src/context.ts
/**
 * Copyright 2025 IBM Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AsyncLocalStorage } from "node:async_hooks";
import { Emitter } from "@/emitter/emitter.js";
import { createRandomHash } from "@/internals/helpers/hash.js";
import { omit } from "remeda";
import { Callback } from "@/emitter/types.js";
import { registerSignals } from "@/internals/helpers/cancellation.js";
import { Serializable } from "@/internals/serializable.js";
import { executeSequentially, LazyPromise } from "@/internals/helpers/promise.js";
import { FrameworkError } from "@/errors.js";
import { shallowCopy } from "@/serializer/utils.js";
import { isAsyncIterable } from "@/internals/helpers/stream.js";

export interface RunInstance<T = any> {
  emitter: Emitter<T>;
}

export interface RunContextCallbacks {
  start: Callback<null>;
  success: Callback<unknown>;
  error: Callback<Error>;
  finish: Callback<null>;
}

export type GetRunContext<T, P = any> = T extends RunInstance ? RunContext<T, P> : never;
export type GetRunInstance<T> = T extends RunInstance<infer P> ? P : never;

export class Run<R, I extends RunInstance, P = any> extends LazyPromise<R> {
  protected readonly tasks: (() => Promise<void>)[] = [];

  constructor(
    handler: () => Promise<R>,
    protected readonly runContext: GetRunContext<I, P>,
  ) {
    super(handler);
  }

  readonly [Symbol.toStringTag] = "Promise";

  observe(fn: (emitter: Emitter<GetRunInstance<I>>) => void) {
    this.tasks.push(async () => fn(this.runContext.emitter));
    return this;
  }

  context(value: object) {
    this.tasks.push(async () => {
      Object.assign(this.runContext.context, value);
      Object.assign(this.runContext.emitter.context, value);
    });
    return this;
  }

  middleware(fn: (context: GetRunContext<I, P>) => void) {
    this.tasks.push(async () => fn(this.runContext));
    return this;
  }

  protected async before(): Promise<void> {
    await super.before();
    await executeSequentially(this.tasks.splice(0, Infinity));
  }

  // @ts-expect-error too complex
  async *[Symbol.asyncIterator](): R extends AsyncIterable<infer X> ? AsyncIterator<X> : never {
    const response = await this.then();
    if (isAsyncIterable(response)) {
      yield* response;
    } else {
      throw new Error("Result is not iterable!");
    }
  }
}

export interface RunContextInput<P> {
  params: P;
  signal?: AbortSignal;
}

export class RunContext<T extends RunInstance, P = any> extends Serializable {
  static #storage = new AsyncLocalStorage<RunContext<any>>();

  protected readonly controller: AbortController;
  public readonly runId: string;
  public readonly groupId: string;
  public readonly parentId?: string;
  public readonly emitter;
  public readonly context: object;
  public readonly runParams: P;
  public readonly createdAt: Date;

  get signal() {
    return this.controller.signal;
  }

  abort(reason?: Error) {
    this.controller.abort(reason);
  }

  constructor(
    public readonly instance: T,
    protected readonly input: RunContextInput<P>,
    parent?: RunContext<any>,
  ) {
    super();
    this.createdAt = new Date();
    this.runParams = input.params;
    this.runId = createRandomHash(5);
    this.parentId = parent?.runId;
    this.groupId = parent?.groupId ?? createRandomHash();
    this.context = omit((parent?.context ?? {}) as any, ["id", "parentId"]);

    this.controller = new AbortController();
    registerSignals(this.controller, [input.signal, parent?.signal]);

    this.emitter = instance.emitter.child<GetRunInstance<T>>({
      context: this.context,
      trace: {
        id: this.groupId,
        runId: this.runId,
        parentRunId: parent?.runId,
      },
    });
    if (parent) {
      this.emitter.pipe(parent.emitter);
    }
  }

  destroy() {
    this.emitter.destroy();
    this.controller.abort(new FrameworkError("Context destroyed."));
  }

  static enter<C2 extends RunInstance, R2, P2>(
    instance: C2,
    input: RunContextInput<P2>,
    fn: (context: GetRunContext<C2, P2>) => Promise<R2>,
  ) {
    const parent = RunContext.#storage.getStore();
    const runContext = new RunContext(instance, input, parent) as GetRunContext<C2, P2>;

    return new Run(async () => {
      const emitter = runContext.emitter.child<RunContextCallbacks>({
        namespace: ["run"],
        creator: runContext,
        context: { internal: true },
      });

      try {
        await emitter.emit("start", null);
        const result = await Promise.race([
          RunContext.#storage.run(runContext, fn, runContext),
          new Promise<never>((_, reject) =>
            runContext.signal.addEventListener("abort", () =>
              setTimeout(() => reject(runContext.signal.reason), 0),
            ),
          ),
        ]);
        await emitter.emit("success", result);
        return result;
      } catch (_e) {
        const e = FrameworkError.ensure(_e);
        await emitter.emit("error", e);
        throw e;
      } finally {
        await emitter.emit("finish", null);
        runContext.destroy();
      }
    }, runContext);
  }

  static {
    this.register();
  }

  createSnapshot() {
    return {
      controller: this.controller,
      runId: this.runId,
      groupId: this.groupId,
      parentId: this.parentId,
      emitter: this.emitter,
      context: shallowCopy(this.context),
      runParams: shallowCopy(this.runParams),
      createdAt: new Date(this.createdAt),
    };
  }

  loadSnapshot(snapshot: ReturnType<typeof this.createSnapshot>) {
    Object.assign(this, snapshot);
  }
}

File: python/beeai_framework/context.py
# Copyright 2025 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import asyncio
import uuid
from collections.abc import Awaitable, Callable
from contextvars import ContextVar
from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Any, Self, TypeVar

from pydantic import BaseModel

from beeai_framework.cancellation import AbortController, AbortSignal, register_signals
from beeai_framework.emitter import Emitter, EmitterInput, EventTrace
from beeai_framework.errors import FrameworkError
from beeai_framework.utils.custom_logger import BeeLogger

R = TypeVar("R")
GetRunContext = TypeVar("GetRunContext", bound="RunContext")

logger = BeeLogger(__name__)


@dataclass
class RunInstance:
    emitter: Emitter


class RunContextInput(BaseModel):
    params: Any
    signal: AbortSignal | None = None


class Run:
    def __init__(self, handler: Callable[[], R], context: GetRunContext) -> None:
        super().__init__()
        self.handler = handler
        self.tasks: list[tuple[Callable[[Emitter], Awaitable[None]], Any]] = []
        self.run_context = context

    def __await__(self) -> R:
        return self._run_tasks().__await__()

    def observe(self, fn: Callable[[Emitter], Awaitable[None]]) -> Self:
        self.tasks.append((fn, self.run_context.emitter))
        return self

    def context(self, context: GetRunContext) -> Self:
        self.tasks.append((self._set_context, context))
        return self

    def middleware(self, fn: Callable[[GetRunContext], None]) -> Self:
        self.tasks.append((fn, self.run_context))
        return self

    async def _run_tasks(self) -> R:
        for fn, param in self.tasks:
            await fn(param)
        self.tasks.clear()
        return await self.handler()

    def _set_context(self, context: GetRunContext) -> None:
        self.context.context = context
        self.context.emitter.context = context


class RunContext(RunInstance):
    storage: ContextVar[Self] = ContextVar("storage", default=None)

    def __init__(
        self, *, instance: RunInstance, context_input: RunContextInput, parent: GetRunContext | None = None
    ) -> None:
        self.instance = instance
        self.context_input = context_input
        self.created_at = datetime.now(tz=UTC)
        self.run_params = context_input.params
        self.run_id = str(uuid.uuid4())
        self.parent_id = parent.run_id if parent else None
        self.group_id = parent.group_id if parent else str(uuid.uuid4())
        self.context = {k: v for k, v in parent.context.items() if k not in ["id", "parentId"]} if parent else {}

        self.emitter = self.instance.emitter.child(
            EmitterInput(
                context=self.context,
                trace=EventTrace(
                    id=self.group_id,
                    run_id=self.run_id,
                    parent_run_id=parent.run_id if parent else None,
                ),
            )
        )

        if parent:
            self.emitter.pipe(parent.emitter)

        self.controller = AbortController()
        parent_signals = [context_input.signal] if parent else []
        register_signals(self.controller, [context_input.signal, *parent_signals])

    @property
    def signal(self) -> AbortSignal:
        return self.controller.signal

    def destroy(self) -> None:
        self.emitter.destroy()
        self.controller.abort("Context destroyed.")

    @staticmethod
    def enter(instance: RunInstance, context_input: RunContextInput, fn: Callable[[GetRunContext], R]) -> Run:
        parent = RunContext.storage.get()
        context = RunContext(instance=instance, context_input=context_input, parent=parent)

        async def handler() -> R:
            emitter = context.emitter.child(
                EmitterInput(namespace=["run"], creator=context, context={"internal": True})
            )

            try:
                await emitter.emit("start", None)

                async def _context_storage_run() -> R:
                    RunContext.storage.set(context)
                    return await fn(context)

                async def _context_signal_aborted() -> str:
                    cancel_future = asyncio.get_event_loop().create_future()

                    def _on_abort() -> None:
                        if not cancel_future.done() and not cancel_future.cancelled():
                            cancel_future.set_result(context.signal.reason)

                    context.signal.add_event_listener(_on_abort)
                    await cancel_future

                abort_task = asyncio.create_task(
                    _context_signal_aborted(),
                    name="abort-task",
                )
                runner_task = asyncio.create_task(_context_storage_run(), name="run-task")

                result = None
                for first_done in asyncio.as_completed([abort_task, runner_task]):
                    result = await first_done
                    abort_task.cancel()
                    break

                await emitter.emit("success", result)
                return result
            except Exception as e:
                error = FrameworkError.ensure(e)
                await emitter.emit("error", error)
                raise
            finally:
                await emitter.emit("finish", None)
                context.destroy()

        return Run(handler, context)
